cmd/snap/color.go

```diff
@@ -0,0 +1,158 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"golang.org/x/crypto/ssh/terminal"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/snap"
+)
+
+type colorMixin struct {
+	Color   string `long:"color" default:"auto" choice:"auto" choice:"never" choice:"always"`
+	Unicode string `long:"unicode" default:"auto" choice:"auto" choice:"never" choice:"always"` // do we want this hidden?
+}
+
+func (mx colorMixin) getEscapes() *escapes {
+	esc := colorTable(mx.Color)
+	if canUnicode(mx.Unicode) {
+		esc.dash = "–" // that's an en dash (so yaml is happy)
+		esc.uparrow = "↑"
+		esc.tick = "✓"
+	} else {
+		esc.dash = "--" // two dashes keeps yaml happy also
+		esc.uparrow = "^"
+		esc.tick = "*"
+	}
+
+	return &esc
+}
+
+func canUnicode(mode string) bool {
+	switch mode {
+	case "always":
+		return true
+	case "never":
+		return false
+	}
+	var lang string
+	for _, k := range []string{"LC_MESSAGES", "LC_ALL", "LANG"} {
+		lang = os.Getenv(k)
+		if lang != "" {
+			break
+		}
+	}
+	if lang == "" {
+		return false
+	}
+	lang = strings.ToUpper(lang)
+	return strings.Contains(lang, "UTF-8") || strings.Contains(lang, "UTF8")
+}
+
+// TODO: maybe unify isTTY (~3 calls just in cmd/snap) (but note stdout vs stdin)
+var isTTY = terminal.IsTerminal(1)
+
+func colorTable(mode string) escapes {
+	switch mode {
+	case "always":
+		return color
+	case "never":
+		return noesc
+	}
+	if !isTTY {
+		return noesc
+	}
+	if _, ok := os.LookupEnv("NO_COLOR"); ok {
+		// from http://no-color.org/:
+		//   command-line software which outputs text with ANSI color added should
+		//   check for the presence of a NO_COLOR environment variable that, when
+		//   present (regardless of its value), prevents the addition of ANSI color.
+		return mono // bold & dim is still ok
+	}
+	if term := os.Getenv("TERM"); term == "xterm-mono" || term == "linux-m" {
+		// these are often used to flag "I don't want to see color" more than "I can't do color"
+		// (if you can't *do* color, `color` and `mono` should produce the same results)
+		return mono
+	}
+	return color
+}
+
+var colorDescs = mixinDescs{
+	"color":   i18n.G("Use a little bit of color to highlight some things."),
+	"unicode": i18n.G("Use a little bit of Unicode to improve legibility."),
+}
+
+type escapes struct {
+	green string
+	end   string
+
+	tick, dash, uparrow string
+}
+
+var (
+	color = escapes{
+		green: "\033[32m",
+		end:   "\033[0m",
+	}
+
+	mono = escapes{
+		green: "\033[1m",
+		end:   "\033[0m",
+	}
+
+	noesc = escapes{}
+)
+
+func (esc *escapes) filler() string {
+	return esc.green + esc.end
+}
+
+func (esc *escapes) longPublisher(storeAccount *snap.StoreAccount) string {
```

Comment:
yes, it feels out of place. I could make them `fooPublisher(*StoreAccount, *escapes)` instead. Would that be better? I'd probably still have them in `color.go` because otherwise where.

Reply:
Yeah, to me this feels better. Fine in color.go, I'm not really fuzzed about the file location (godef will find it for me ;).